### PR TITLE
[FLINK-14935][task,runtime] Use RunnableWithException in the Mailbox

### PR DESCRIPTION
**Pull request by pnowojski**
This allows mailbox action to throw an exception.

## Verifying this change

This change is already covered by existing tests and adds a new one `TaskMailboxProcessorTest#testSubmittingRunnableWithException` 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/MailboxExecutor.java [hunk 1]
```diff
@@ -227,21 +226,4 @@ default void execute(@Nonnull Runnable command, String description) {
...
-	 * <p>An optional description can (and should) be added to ease debugging and error-reporting. The description
-	 * may contain placeholder that refer to the provided description arguments using {@link java.util.Formatter}
-	 * syntax. The actual description is only formatted on demand.
-	 *
-	 * @param descriptionFormat the optional description for all commands that is used for debugging and
-	 * error-reporting.
-	 * @param descriptionArgs the parameters used to format the final description string.
-	 * @return an {@code Executor} view on this {@code MailboxExecutor}
-	 */
-	default Executor asExecutor(String descriptionFormat, Object... descriptionArgs) {
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/MailboxExecutor.java]: FYI, having a bridge to java...Executor was meant to help implementing CompletableFutures. However, since the interface is so simple, it might as well be a lambda implementation when it's actually used in CompletableFutures.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java [hunk 2]
```diff
@@ -68,14 +70,23 @@ public void executeFirst(
...
 
 	@Override
 	public boolean tryYield() {
 		Optional<Mail> optionalMail = mailbox.tryTake(priority);
 		if (optionalMail.isPresent()) {
-			optionalMail.get().run();
+			try {
+				optionalMail.get().run();
+			} catch (Exception ex) {
+				throw new FlinkRuntimeException(ex);
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java]: Please use `WrappingRuntimeException.wrapIfNecessary`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java [hunk 3]
```diff
@@ -68,14 +70,23 @@ public void executeFirst(
 
 	@Override
 	public void yield() throws InterruptedException {
-		mailbox.take(priority).run();
+		Mail mail = mailbox.take(priority);
+		try {
+			mail.run();
+		} catch (Exception ex) {
+			throw new FlinkRuntimeException(ex);
```
Comment (rkhachatryan) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java]: Does wrapping into FlinkRuntimeException provide any additional info compared to RuntimeException?

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Adapter from {@link FutureTask} to {@link RunnableWithException}.
+ */
+public class FutureTaskWithException<V> implements Future<V>, RunnableWithException {
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]: Can't we just use FutureTask and wrap the RunnableWithException into a Callable as you did?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java [hunk 5]
```diff
@@ -61,11 +62,7 @@ public String toString() {
 		return String.format(descriptionFormat, descriptionArgs);
 	}
 
-	public void run() {
-		try {
-			runnable.run();
-		} catch (Exception e) {
-			throw new IllegalStateException("Cannot process mail " + toString(), e);
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java]: Why did you move exception handling out? Now we have 3 places instead of 1 where we do the exception translation.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/MailboxExecutor.java (hunk 1 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/MailboxExecutor.java]: Looking at the last commit, will actually make the mentioned adhoc implementation quite verbose. If you think CompleteableFutures over mailbox is a valid use case, reconsider removing this method.

Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/MailboxExecutor.java]: It's not used currently anywhere, so I prefered to drop it instead of implementing `Runnable` -> `RunnableWithException` wrapping of the `Executor` interface. Now I see that that's only method (previously I thought it's more), but still - I would prefer to not maintain a dead code. We can re-introduce it later if there will be such demand.

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java (hunk 4 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]: Unfortunately not. I tried to do just that (actually it was even simpler, since it was enough to pass a `futureTask::run` reference, without any wrapping to convert it into a `RunnableWithException`), but it didn't work since there are `instanceof Future` checks inside the `FutureUtils#cancelRunnableFutures` called by `MailboxProcessor#close`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java (hunk 5 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java]: I haven't moved it out - I just dropped this wrapping and now I'm letting exception to bubble up through the `MailboxProcessor#runMailboxLoop` callstack. It only needed to be wrapped into a runtime exception inside `yield` and `tryYield` methods to not pollute user's code.

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+	@Override
+	public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+		return futureTask.get(timeout, unit);
+	}
+
+	@Override
+	public void run() throws Exception {
+		futureTask.run();
+	}
```
Comment (rkhachatryan) [inline flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]: If an exception is thrown inside futureTask.run() then it is set as futureTask outcome, which is ignored in this method, so effectively it gets swallowed.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java (hunk 5 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java]: That's exactly what I meant by moving exception handling out. Now we have 3 places instead of 1 where we need to wrap it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java (hunk 4 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]: I didn't get your answer, so let's work with code:

```java
	default @Nonnull Future<Void> submit(@Nonnull RunnableWithException command, String description) {
		FutureTask<Void> future = new FutureTask<>(asCallable(command, null));
		execute(future, description, EMPTY_ARGS);
		return future;
	}
```

Where asCallable is your function from FunctionUtils.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java [hunk 7]
```diff
@@ -68,14 +70,23 @@ public void executeFirst(
...
 	@Override
 	public void yield() throws InterruptedException {
-		mailbox.take(priority).run();
+		Mail mail = mailbox.take(priority);
+		try {
+			mail.run();
+		} catch (Exception ex) {
+			throw WrappingRuntimeException.wrapIfNecessary(ex);
+		}
 	}
```
Comment (rkhachatryan) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java]: We execute user-supplied code here and call it e.g. from StreamTask.triggerCheckpointOnBarrier, so probably it's better not to hide the exception but to declare it or at least document with @ throws.

File: flink-core/src/main/java/org/apache/flink/util/function/FunctionUtils.java [hunk 8]
```diff
@@ -115,4 +116,14 @@ private FunctionUtils() {
 			return result;
 		};
 	}
+
+	/**
+	 * Converts a {@link RunnableWithException} into a {@link Callable} that will return the {@code result}.
+	 */
+	public static <T> Callable<T> asCallable(RunnableWithException command, T result) {
```
Comment (AHeise) [inline flink-core/src/main/java/org/apache/flink/util/function/FunctionUtils.java]: Could we move that static function to a default member function of RunnableWithException?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java (hunk 5 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java]: Why 3? I see just `yield` and `tryYield`.

Besides that not wrapping the exception `MailboxProcessor#runMailboxLoop` is more or less whole point of this change/feature.

File: flink-core/src/main/java/org/apache/flink/util/function/FunctionUtils.java (hunk 8 above)
Comment (pnowojski) [reply flink-core/src/main/java/org/apache/flink/util/function/FunctionUtils.java]: `RunnableWithException` is `@Public`, while `FunctionUtils` are not. That's why I placed it here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java (hunk 4 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]:
```
execute(future, description, EMPTY_ARGS);
```
Compile error
```
Error:(166, 25) java: incompatible types: java.util.concurrent.FutureTask<java.lang.Void> cannot be converted to org.apache.flink.util.function.RunnableWithException
```
Change to:
```
execute(future::run, description, EMPTY_ARGS);
```
results with test failures, since this submitted `RunnableWithException` is no longer a `Future`, so it's not being canceled on `close()`

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java (hunk 6 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]: Isn't this an expected behaviour? it's used in for example 
```
Future<T> MailboxExecutor#submit(java.util.concurrent.Callable<T>, java.lang.String)
```
which is supposed to report a failure via the `Future`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java (hunk 7 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java]: But the designed use case of `yield` is inside user's operators. There we don't want to pollute user code with some random exceptions that he can not handle in any way (and we would force him to either declare `throws Exception` everywhere in his code, or to wrap it into the runtime anyway).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java (hunk 3 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java]: A little bit - it allows us/user to classify runtime exceptions that originate from Flink. Also this is I think our convention and potentially a bigger discussion whether we need `FlinkException`/`FlinkRuntimeException` hierarchy.

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java (hunk 4 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]: My example did compile for me. Please double-check your setup. Or give me more context. Not sure why it would convert the `FutureTask` to `RunnableWithException`.

Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]: Btw internally, FutureTask just uses Callables and wraps Runnable into Callable. I'm just proposing to do the same on call-site.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java (hunk 5 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java]: Oh, I missed that, then it's even worse. The whole point of the exception translation is to actually use the the `#toString()` of Mail to give some context information about the mail being processed for failure debugging.

File: flink-core/src/main/java/org/apache/flink/util/function/FunctionUtils.java (hunk 8 above)
Comment (AHeise) [reply flink-core/src/main/java/org/apache/flink/util/function/FunctionUtils.java]: Even more reason to add something that is useful for the all users. ;)

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java (hunk 6 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]: You are right. 
I was misguided by `throws Exception` - should we remove it then?

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java (hunk 4 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]: resolved after an offline discussion: we decided that `FutureTaskWithException` will stay, but it can be simplified.

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java (hunk 6 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureTaskWithException.java]: Yes, this `throws Exception` could have been removed. But whole class will be simplified and this method will go away completely anyway :)

File: flink-core/src/main/java/org/apache/flink/util/function/FunctionUtils.java (hunk 8 above)
Comment (pnowojski) [reply flink-core/src/main/java/org/apache/flink/util/function/FunctionUtils.java]: ok, I will move it to the `RunnableWithException`.

**Comment by pnowojski** [comment]
@AHeise could you take a look at the updated code and give more comments +1 the code?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java (hunk 3 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java]: We don't know here whether the exception originated from user-code or internal Flink code.

Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java]: That's right about the "cause" of this exception that we are throwing here. But this exception is being thrown by us.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java (hunk 5 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java]: This comment has not been addressed, but is important to me. It reverts some changes done in FLINK-14199 for improved error reporting.

Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mail.java]: Ok, I see.

Btw, clearly this try/catch and `toString` was missing a test coverage
